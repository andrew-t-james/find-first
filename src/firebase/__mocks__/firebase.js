const mockUser = {
  user: {
    uid: 1,
    displayName: 'Steve',
    photoURL: 'some-url'
  }
};

export const initializeApp = jest.fn()
  .mockImplementation(() => {});

// export const firebase = {
//   auth: () => ({
//     GoogleAuthProvider: jest.fn(() => {}),
//     signInWithPopup: () => new Promise(resolve => resolve(mockUser)),
//     signOut: () => jest.fn()
//   })
// };

export const logout = jest.fn();
export const googleOAuthLogin = () => new Promise(resolve => resolve(mockUser));
export const githubOAuthLogin = () => new Promise(resolve => resolve(mockUser));