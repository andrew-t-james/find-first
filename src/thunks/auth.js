export const googleLogin = (authProvider, loginAction) => {
  return async dispatch => {
    const newUser = await authProvider();
    dispatch(loginAction(newUser.user));
  };
};

export const googleLogout = (authLogout, logoutAction) => {
  return async dispatch => {
    await authLogout();
    dispatch(logoutAction());
  };
};