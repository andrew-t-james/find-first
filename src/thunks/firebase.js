import { isLoadingAction } from '../Actions/loader';
import database from '../firebase/firebase';
import { saveJobAction } from '../Actions/saved-jobs';

export const addJobToFirebase = newJob => {
  return async dispatch => {
    dispatch(isLoadingAction(true));

    await database.ref('saved-jobs').push(newJob);
    dispatch(saveJobAction({
      ...newJob
    }));

    dispatch(isLoadingAction(false));
  };
};