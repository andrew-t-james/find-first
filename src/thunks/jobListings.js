import { githubJobsAction } from '../Actions/github';
import { githubApiRequest } from '../helpers/api-helpers';
import { isLoadingAction } from '../Actions/loader';

export const githubJobsThunk = () => {
  return async dispatch => {
    dispatch(isLoadingAction(true));
    try {
      const githubJobsList = await githubApiRequest();
      dispatch(isLoadingAction(false));
      dispatch(githubJobsAction(githubJobsList));
    } catch (error) {
      throw Error(error.message);
    }
  };
};