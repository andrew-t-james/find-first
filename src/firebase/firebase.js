import * as firebase from 'firebase';

const config = {
  apiKey: process.env.FIREBASE_API_KEY,
  authDomain: process.env.FIREBASE_AUTH_DOMAIN,
  databaseURL: process.env.FIREBASE_DATABASE_URL,
  projectId: process.env.FIREBASE_PROJECT_ID,
  storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.FIREBASE_MESSAGING_SENDER_ID
};

firebase.initializeApp(config);

export const googleAuthProvider = new firebase.auth.GoogleAuthProvider();
export const githubAuthProvider = new firebase.auth.GithubAuthProvider();
export const twitterAuthProvider = new firebase.auth.TwitterAuthProvider();
export const facebookAuthProvider = new firebase.auth.FacebookAuthProvider();

export const googleOAuthLogin = () => firebase.auth().signInWithPopup(googleAuthProvider);
export const githubOAuthLogin = () => firebase.auth().signInWithPopup(githubAuthProvider);
export const twitterOAuthLogin = () => firebase.auth().signInWithPopup(twitterAuthProvider);
export const facebookOAuthLogin = () => firebase.auth().signInWithPopup(facebookAuthProvider);
export const logout = () => firebase.auth().signOut();

const database = firebase.database();

export { firebase, database as default };


// * how to push into firebase
// database.ref('notes').push({
//   title: 'todo',
//   body: 'go for a run'
// });

// * editing array data in firebase
// database.ref('notes/-LI6UR9CVl1FaV46zinX').update({
//   title: 'dont todo',
//   body: 'yeah not doing it'
// });



// database.ref('notes')
//   .once('value')
//   .then(snapshot => {
//     const notes = [];
//     snapshot.forEach(snap => {
//       notes.push({
//         id: snap.key,
//         ...snap.val()
//       });
//     });
//     console.log(notes);
//   });


// database.ref('notes').on('value', snapshot => {
//   const expenses = [];

//   snapshot.forEach((childSnapshot) => {
//     expenses.push({
//       id: childSnapshot.key,
//       ...childSnapshot.val()
//     });
//   });

//   console.log(expenses);
// });

// database.ref('notes').on('child_removed', snapshot => {
//   console.log(snapshot.key, snapshot.val());
// });


// database.ref('notes')
//   .once('value')
//   .then((snapshot) => {
//     const expenses = [];

//     snapshot.forEach((childSnapshot) => {
//       expenses.push({
//         id: childSnapshot.key,
//         ...childSnapshot.val()
//       });
//     });

//     console.log(expenses);
//   });