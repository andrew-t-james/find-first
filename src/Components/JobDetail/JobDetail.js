import React from 'react';
import { Header } from '../Header/Header';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { addJobToFirebase } from '../../thunks/firebase';

export const JobDetail = ({ title, description, url, image, saveJob }) => {
  const newJob = {
    title,
    description,
    url,
    image
  };

  return (
    <section className="job-detail">
      <header className="job-detail__header">
        <h1>header logo goes here</h1>
      </header>
      <section className="job-detail-container">
        <div className="job-detail--buttons">
          <Link to="/dashboard">&larr; Back to Job Listings</Link>
          <img src={image} alt={title} className="job-detail__image"/>
          <a href={url} target="_blank" rel="noopener noreferrer" >Apply Here &rarr;</a>
        </div>
        <h1 className="job-detail__tile">{title}</h1>
        <div className="job-detail__save-job">
          <button onClick={() => saveJob(newJob)}
            className="job-detail__save-job--button"
          >
            Mark Applied
          </button>
        </div>
        <p className="job-detail__copy" dangerouslySetInnerHTML={{ __html: description }}/>
      </section>
    </section>
  );
};

export const mapDispatchToProps = dispatch => ({
  saveJob: newJob => dispatch(addJobToFirebase(newJob))
});

export default connect(null, mapDispatchToProps)(JobDetail);