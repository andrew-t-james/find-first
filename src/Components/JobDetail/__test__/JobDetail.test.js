import React from 'react';
import { shallow } from 'enzyme';
import { JobDetail, mapDispatchToProps, mapStateToProps } from '../JobDetail';
import { addJobToFirebase } from '../../../thunks/firebase';

jest.mock('../../../firebase/firebase.js');
jest.useFakeTimers();

describe('<JobDetail />', () => {
  let wrapper;
  const mockJobListing = [
    {
      id: 1,
      title: 'Title',
      description: 'description'
    },
    {
      id: 2,
      title: 'Title',
      description: 'description'
    }
  ];

  const mockSavedJobs = [...mockJobListing];
  const mockSaveJob = jest.fn();

  beforeEach(() => wrapper = shallow(
    <JobDetail
      savedJobs={mockSavedJobs}
      saveJobToFirebase={mockSaveJob}
      {...mockJobListing}
    />
  ));

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call saveNewJob on click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'saveNewJob');
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find('.job-detail__save-job--button').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('should updateState', () => {
    const expected = {
      saving: false,
      savedToDB: true
    };
    const mockJobToSave = {
      id: 2,
      title: 'title'
    };
    wrapper.instance().saveNewJob(mockJobToSave);
    jest.runAllTimers();
    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1300);
    setTimeout(()=>{
      expect(wrapper.state('saving')).toBe(false);
      expect(wrapper.state('savedToDB')).toBe(true);
    }, 1500);
  });


  describe('mapStateToProps', () => {
    test('should map savedJobs to props', () => {
      const expected = {
        savedJobs: mockSavedJobs
      };
      const mockState = {
        savedJobs: mockSavedJobs
      };
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    });

  });

  describe('mapDispatchToProps', () => {
    test('should call dispatch with addJobToFirebase', () => {
      const mockDispatch = jest.fn();
      const mockJob = {
        title: 'title',
        description: 'description',
        image: 'some-url-here',
        url: 'another-url-here'
      };
      const actionToDispatch = addJobToFirebase(mockJob);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.saveJobToFirebase(mockJob);

      expect(mockDispatch).toHaveBeenCalled();
    });
  });
});
