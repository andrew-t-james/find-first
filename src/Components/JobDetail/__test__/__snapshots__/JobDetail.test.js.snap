// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JobDetail /> should match snapshot 1`] = `
<section
  className="job-detail"
>
  <header
    className="job-detail__header"
  >
    <div
      className="header__logo"
    >
      <h1
        className="logo__forward"
      >
        F
      </h1>
      <h1
        className="logo__backward"
      >
        F
      </h1>
    </div>
  </header>
  <section
    className="job-detail-container"
  >
    <div
      className="job-detail--buttons"
    >
      <Link
        replace={false}
        to="/dashboard"
      >
        ← Back to Job Listings
      </Link>
      <div />
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        Apply Here →
      </a>
    </div>
    <h1
      className="job-detail__tile"
    />
    <div
      className="job-detail__save-job"
    >
      <button
        className="job-detail__save-job--button
              
              "
        disabled={false}
        onClick={[Function]}
      >
        Mark Applied
      </button>
    </div>
    <p
      className="job-detail__copy"
      dangerouslySetInnerHTML={
        Object {
          "__html": undefined,
        }
      }
    />
  </section>
</section>
`;
